packages/donut-monster/test/index.test.js

```diff
@@ -0,0 +1,77 @@
+import getPort from 'get-port'
+import socketIoClient from 'socket.io-client'
+import test from 'ava'
+import { getRandomDonuts, messages } from 'donut-common'
+import { promisify } from 'util'
+
+import io from '../src/index.js'
+
+const getClient = () => socketIoClient(`http://localhost:${io.httpServer.address().port}`)
+
+test.before(async () => {
+  const port = await getPort()
+  return promisify(io.httpServer.listen.bind(io.httpServer))(port)
+})
+
+test.serial('initial connection state', (t) => new Promise((resolve, reject) => {
+  const client = getClient()
+
+  client.on(messages.INIT_CLIENT, ({ submitMode }) => {
+    t.false(submitMode, 'submit mode initially off')
+    resolve()
+  })
+  client.on('disconnect', reject)
+  client.on('error', reject)
+}))
+
+// Test is stateful
+test.serial('submit mode', (t) => new Promise((resolve, reject) => {
+  const client = getClient()
+
+  client.on(messages.SUBMIT_MODE, (submitMode) => {
+    t.true(submitMode, 'sets submit mode')
+    resolve()
+  })
+  client.emit(messages.SUBMIT_MODE, true)
+
+  client.on('disconnect', reject)
+  client.on('error', reject)
+}))
+
+// Actually computes regression, this one takes a while
+test.serial('uploads donuts', (t) => new Promise((resolve, reject) => {
+  const client = getClient()
+  const donuts = getRandomDonuts(10)
+  const resolveSometime = () => {
+    if (++resolveSometime.count > 2) resolve()
```

Comment:
`t.plan` doesn’t work (I think the event queue is drained?). Gotta wrap with a `new Promise` and do this weirdness.